### PR TITLE
feat: pass req object to authenticate method

### PR DESCRIPTION
**Pull request by itamar-owlytics**

### CODE REVIEW

File: src/authentication/login.handler.ts [hunk 1]
```diff
@@ -89,7 +89,7 @@ export const withLogin = (
       email: string;
       password: string;
     };
-    const adminUser = await auth.authenticate(email, password);
+    const adminUser = await auth.authenticate(email, password, req);
```
Comment (plorencrstit) [inline src/authentication/login.handler.ts:92]: Why do you need it?

Comment (itamar-owlytics) [reply src/authentication/login.handler.ts:92]: Because I want to get the IP address of the authenticated user

Comment (dziraf) [reply src/authentication/login.handler.ts:92]: Why would you need the IP address in your `authenticate` method? This looks like a very rare case where you'd probably end up better creating your own authentication mechanism.

Comment (trangvu9) [reply src/authentication/login.handler.ts:92]: hey how do we create custom authentication?

Comment (itamar-owlytics) [reply src/authentication/login.handler.ts:92]: > Why would you need the IP address in your `authenticate` method? This looks like a very rare case where you'd probably end up better creating your own authentication mechanism

I want to save the IP address of the authenticated user and maybe in the future I will want to block authentications from unknown IPs

**Comment by itamar-owlytics** [comment]
@plorencrstit @dziraf 
What do you say?

**Comment by trangvu9** [comment]
Say yes say yes ;)) 

No kidding, this change opens a whole lots of possibilities, you can take a sneak peak of the cookie as well, in case you put the whole thing behind a auth wall. 

Also I don’t see anywhere it might hurt!

**Comment by dziraf** [comment]
While `Request` can be useful for authenticating in case of custom authentication mechanisms, so far if we ever had to access the `Request`, we implemented our custom `POST /login` controller. Adding `req` as the third argument can fulfill your requirements but it is also not the cleanest way. I plan to rewrite all framework plugins so that they use a unified interface and are more customizable, especially for authentication, but until then I suggest that you just implement a custom `buildAuthenticatedRouter` based on this: https://github.com/SoftwareBrothers/adminjs-expressjs/blob/master/src/buildAuthenticatedRouter.ts

**Pull request closed without merging**